[Woo POS] M2: Add modal banner for simple products only

soundsokay opened the pull request:
The PR adds a new modal view, `SimpleProductsModalView`, that provides details about only simple products being supported by Woo POS.

The new view may be triggered from the existing header banner that states "Showing simple products only" when POS mode is entered (see #13317).

Closing the header banner causes an info icon to be shown in the Products header. Tapping that will display the new modal banner for simple products.

The PR also centralizes modal presentation logic in PointOfSaleDashboardViewModel (commit 1f4084b3e9a1ea31096b73ec121555ca54e2013c).

The new view doesn't yet match the design in a pixel-perfect manner per prior agreement this week. The link on the modal view to "Create an order in store management" also doesn't yet work. These can be taken up under a separate ticket.

<!-- Remember about a good descriptive title. -->

Closes: #13357
<!-- Id number of the GitHub issue this PR addresses. -->


## Testing information

- Enter POS mode.
- Observe the "Showing simple products only" banner is shown.
- Tap the banner or the banner's info icon.
-  Observe the modal is show with "Why can't I see my products?" as the title.
- Click OK or the X icon to close the modal.
- Click the X icon in the header banner.
- Observe an info icon appears in the Products header.
- Tap the info icon.
- Observe the "Why can't I see my products?" modal is shown.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/3edd4439-b633-46ee-aa50-c4e2e5703f95





---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

soundsokay committed:
feat: adds modal banner for simple products only

- adds SimpleProductsModalView.
- updates ItemListViewModel and adds a unit test.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

bozidarsevo commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -5,30 +5,44 @@ struct ItemListView: View {
     @ScaledMetric private var scale: CGFloat = 1.0
     @ObservedObject var viewModel: ItemListViewModel
     @Environment(\.floatingControlAreaSize) var floatingControlAreaSize: CGSize
+    @State private var showModal: Bool = false
```
Maybe `showSimpleProductsModal`? Just in case we add more/different modals.

bozidarsevo commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -47,23 +61,22 @@ private extension ItemListView {
             HStack {
                 headerTextView
                 Spacer()
-                Button(action: {
-                    // TODO:
-                    // https://github.com/woocommerce/woocommerce-ios/issues/13357
-                    debugPrint("Not implemented")
-                }, label: {
-                    Image(uiImage: .infoImage)
-                })
```
Why is this removed? There should be an info button when that initial banner is removed.
In https://github.com/woocommerce/woocommerce-ios/pull/13417/files I started updating product list and left a method `openInfoBanner()` that could call `showModal.toggle()` and we can call it from this button and when tapping the whole banner card

<img width="385" alt="Screenshot 2024-07-25 at 10 18 52" src="https://github.com/user-attachments/assets/2cb2707f-02ce-437c-a161-1b08cc4caf6d">

soundsokay replied:
I added it back. Looks like we both worked on the same code. I incorporated your styling code into mine during resolving merge conflicts.

bozidarsevo commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -5,30 +5,44 @@ struct ItemListView: View {
+            .padding(.horizontal, Constants.itemListPadding)
+            .background(Color.posBackgroundGreyi3)
+
+            if showModal {
+                SimpleProductsModalView(isPresented: $showModal)
+                    .background(Color.clear)
+                    .transition(.move(edge: .bottom))
+                    .onAppear {
+                        withAnimation(.easeInOut) {
+                            showModal = true
+                        }
+                    }
```
I suppose we need to show this one and other modals above the main ZStack in the POS, because now it is showing on top of products and messes up the layout. We could use one of our view models to set which modal should be shown on top of entire POS UI

![Simulator Screenshot - iPad Pro (11-inch) (4th generation) 17 2 - 2024-07-25 at 10 22 15](https://github.com/user-attachments/assets/6dff3c04-cf08-4dab-aff0-45c52f6ef6f9)

<img width="608" alt="Screenshot 2024-07-25 at 10 22 12" src="https://github.com/user-attachments/assets/718df306-0659-474b-885b-8ed18689c936">

soundsokay replied:
This has been addressed.

soundsokay replied:
> We could use one of our view models to set which modal should be shown on top of entire POS UI

Good idea. Implemented in commit 1f4084b3e9a1ea31096b73ec121555ca54e2013c.

soundsokay replied:
I needed to back this out and re-merge trunk. An earlier merge of trunk led to a tangle that I thought might need generics to sort out.

Suggest we can take up centralizing modal view presentation in a later PR.

soundsokay replied:
>We could use one of our view models to set which modal should be shown on top of entire POS UI

I added this back in commit a5df2a7be8514ddbc2d37972a7712304d2321f1d

soundsokay committed:
refactor: rename 'showModal' to 'showSimpleProductsModal'

- Addresses feedback that the new name is more clear especially in case different modals are added later.

soundsokay committed:
refactor: Add and style info button in ItemListView

- Reintroduced the info button in the header when the initial banner is dismissed.
- Adjusted the size of the info button to ensure it does not overshadow the "Products" text.

soundsokay committed:
fix: Hide info button at app launch until header banner is dismissed

- Adjusted the visibility condition for the info button in the header to ensure it only appears when the header banner is dismissed.
- This change ensures the info button does not incorrectly appear when POS mode is entered.

soundsokay committed:
feat: show SimpleProductsModalView properly centered on top without disrupting the view hierarchy

- Present the modal from PointOfSaleDashboardView.

soundsokay committed:
refactor: remove unused code

soundsokay committed:
feat: make header banner tappable

- Displays modal banner: SimpleProductsModalView
- Addresses PR feedback

soundsokay committed:
feat: align SimpleProductsModalView with the design

- Adds private extension for constants and localization.

soundsokay committed:
chore: fix linting issue and remove comments

- Remove whitespace.

soundsokay committed:
chore: fix linting issue

soundsokay committed:
Merge remote-tracking branch 'origin/trunk' into feat/simple-products-only-banner-13357

Conflicts:
	WooCommerce/Classes/POS/Presentation/ItemListView.swift

soundsokay committed:
refactor: Centralize modal presentation logic in PointOfSaleDashboardViewModel

- Added `ModalType` enum to `PointOfSaleDashboardViewModel` to manage modal types
- Conformed `ModalType` enum to `Identifiable` to support SwiftUI `sheet` modifier
- Updated `ItemListView` to present modals using `dashboardViewModel.activeModal`
- Refactored `SimpleProductsModalView` to use localized strings and constants
- Consolidated modal presentation logic to ensure only one modal is shown on top of the entire POS UI

Addresses reviewer feedback to improve the maintainability and scalability of modal management in the POS UI.

soundsokay committed:
chore: fix linting issues

soundsokay committed:
test: add test coverage for showModal, hideModal, and ModalType conformance to Identifiable

bozidarsevo commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -36,6 +36,22 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
+
+        var id: String {
+            switch self {
+            case .simpleProducts:
```
Good thing there is no default case. When we add new cases the compiler will show us where we need to add some logic :) p2: pfoUAQ-gM-p2

soundsokay committed:
Merge branch 'trunk' into feat/simple-products-only-banner-13357-fix-linting-f1cda2-merge-trunk

Conflicts:
WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift
WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift

soundsokay committed:
feat: refactor SimpleProductsModalView integration and update UI overlay structure

This commit ensures that the SimpleProductsModalView is properly integrated within the PointOfSaleDashboardView, with the floating toolbar items correctly displayed beneath the modal overlay.

soundsokay committed:
refactor: restore centralized modal presentation logic

- Second implementation of this which was first added in commit 1f4084b: refactor: Centralize modal presentation logic in PointOfSaleDashboardViewModel.
- Add test coverage in PointOfSaleDashboardViewModelTests.

These changes aim to streamline modal presentation logic and improve code maintainability.

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -37,6 +37,17 @@ struct PointOfSaleDashboardView: View {
+                    .zIndex(1)
+            }
+
+            if viewModel.shouldShowModal(.simpleProducts) {
```
I read the discussion with @bozidarsevo but I couldn't understand 100% why we cannot use native sheet presentation logic? 

It will also streamline some of the design work (modal size, rounded corners, titles), rather than showing a custom view on top of a `ZStack`?

If I understood correctly, the concern may have been that we show pop up on top of product list, rather than on top of the whole pos but we could implement the logic in the dashboard view and do the same as with the card reader pop-up:
`.sheet(isPresented: $totalsViewModel.showsCardReaderSheet`?

Are there any blockers around this?

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -37,6 +37,17 @@ struct PointOfSaleDashboardView: View {
+
+            if viewModel.shouldShowModal(.simpleProducts) {
+                SimpleProductsModalView(isPresented: Binding(
+                    get: { viewModel.shouldShowModal(.simpleProducts) },
```
May be enough to simplify it to `viewModel.showsSimpleProductsSheet` if we use the native approach?

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -130,7 +136,7 @@ private extension PointOfSaleDashboardView {
 
     var productListView: some View {
-        ItemListView(viewModel: itemListViewModel)
+        ItemListView(viewModel: self.itemListViewModel, dashboardViewModel: viewModel)
```
Although we haven't yet found a good approach for our architecture (https://github.com/woocommerce/woocommerce-ios/issues/13441), maybe we could pass [a binding](https://developer.apple.com/documentation/swiftui/binding#overview) here rather than a whole dashboardViewModel? 

The idea being that `ItemListView` could change `isShowing = true` and then `PointOfSaleDashboardView` would react accordingly.

staskus reviewed (commented):
Thanks for the work!

My main question comes from the decision to implement custom modals rather than relying on native sheets.  As I understand it comes from the previous discussion but I want to make sure I understand the reasoning.

Plus, modals are managed by ItemListViewModel while shown on PointOfSaleDashboardView so there may be an opportunity to sort out the dependencies.

staskus commented:
Based on modal work in other places https://github.com/woocommerce/woocommerce-ios/pull/13466#discussion_r1698732237 I understand that native sheets do not satisfy all our requirements.

joshheald commented:
@soundsokay I'm taking a look at this issue, as we have some new modal code, so I'm going to make it work with that... hope that's OK. I'll close this PR for clarity.
